tortoise/verifying.go

```diff
@@ -149,39 +170,43 @@ func (v *verifying) verify(logger log.Log, lid types.LayerID) bool {
 		log.Stringer("local_threshold", v.localThreshold),
 		log.Stringer("global_threshold", v.globalThreshold),
 	)
-
-	// 0 - there is not enough weight to cross threshold.
-	// 1 - layer is verified and contextual validity is according to our local opinion.
 	if sign(margin.Cmp(v.globalThreshold)) == abstain {
-		logger.With().Info("candidate layer is not verified." +
-			" voting weight from good ballots is lower than the threshold")
+		logger.With().Debug("doesn't cross global threshold")
 		return false
 	}
-
-	// if there is any block with neutral local opinion - we can't verify the layer
-	// if it happens outside hdist - protocol will switch to full tortoise
-	for _, bid := range v.blocks[lid] {
-		vote, _ := getLocalVote(v.commonState, v.Config, lid, bid)
-		if vote == abstain {
-			logger.With().Info("candidate layer is not verified."+
-				" block is undecided according to local votes",
-				bid,
-			)
-			return false
-		}
-		v.validity[bid] = vote
+	if verifyLayer(
+		logger,
+		v.blocks[lid],
+		v.validity,
+		func(block blockInfo) sign {
+			decision, _ := getLocalVote(v.commonState, v.Config, lid, block.id)
```

Comment:
the code as is still votes abstain for zdist layers even when hare outputs empty layer immediately. is that deliberate or should that be a TODO?

Reply:
are you referring to getLocalVote internals? no, it doesn't vote abstain if hare output for the layer is finalized